### PR TITLE
Lazily enable zpages at runtime

### PR DESCRIPTION
**Pull request by ramonza**
See: #772

### CODE REVIEW

File: zpages/zpages.go [hunk 1]
```diff
@@ -31,16 +31,40 @@
...
 
 import (
+	"fmt"
+	"log"
+	"net"
 	"net/http"
 )
 
-// Handler is an http.Handler that serves the zpages.
+// Handler is deprecated: Use NewHandler.
```
Comment (ramonza) [inline zpages/zpages.go]: Once this is removed according to the deprecation policy, we can move the init logic into NewHandler.

File: examples/grpc/helloworld_server/main.go [hunk 2]
```diff
@@ -31,6 +30,7 @@ import (
 	"go.opencensus.io/zpages"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
+	"net/http"
```
Comment (rakyll) [inline examples/grpc/helloworld_server/main.go]: Revert this change.

File: zpages/zpages.go [hunk 3]
```diff
@@ -31,16 +31,31 @@
...
 var Handler http.Handler
 
+var enableOnce sync.Once
+
 func init() {
+	Handler = NewHandler("")
+}
+
+// GetHandler returns a handler that serves the z-pages.
```
Comment (rakyll) [inline zpages/zpages.go]: NewHandler

File: zpages/zpages.go [hunk 4]
```diff
@@ -31,16 +31,31 @@
...
+func NewHandler(prefix string) http.Handler {
```
Comment (rakyll) [inline zpages/zpages.go]: We shouldn't have the prefix here.

It is coupling the responsibilities of a mux and a handler into a handler.

Comment (ramonza) [reply zpages/zpages.go]: I this this is preferable to using StripPrefix. It is much more readable, does not allocate a new request object, and does not try to fake the URL (like StripPrefix does).

**Comment by ramonza** [comment]
Thanks @rakyll updated. I prefer the approach with taking a prefix as explained above and would like to keep it unless you feel strongly that it's wrong. After looking at what StripPrefix does, I don't think it's a good idea to recommend its use here if there's an straightforward way to avoid it.

**Comment by ramonza** [comment]
So one way to fix this would be to not care what the prefix is and just match based on the last component of the URL. What do you think about that? Something like:

```go
parts := path.Split(req.URL)
page := parts[len(parts-1)]
switch page {
case "rpcz": 
...
default:
return http.NotFoundHandler().Handle(...)
}
```

Then we wouldn't need to take a prefix parameter at all and it would work regardless of the path it is installed at?

File: zpages/zpages.go [hunk 5]
```diff
@@ -31,16 +31,31 @@
...
+	"fmt"
 	"net/http"
+	"sync"
+
+	"go.opencensus.io/internal"
 )
 
-// Handler is an http.Handler that serves the zpages.
+// Handler is deprecated: Use NewHandler.
 var Handler http.Handler
```
Comment (nfisher) [inline zpages/zpages.go]: If this were changed to http.ServerMux I think that would work for what I referenced in #783. The http.Handler strips the `.Handle*` functions and while I could simply cast this to a ServerMux it feels dirty as I'm relying on knowledge of the implementation detail rather than a public contract.

File: zpages/zpages.go [hunk 6]
```diff
@@ -31,16 +31,32 @@
...
-	zpagesMux.HandleFunc("/tracez", tracezHandler)
-	zpagesMux.Handle("/public/", http.FileServer(fs))
-	Handler = zpagesMux
+	mux := http.NewServeMux()
+	AppendHandlers("", mux)
+	Handler = mux
+}
+
+// AppendHandlers adds the z-pages to the given ServeMux rooted at pathPrefix.
+func AppendHandlers(pathPrefix string, mux *http.ServeMux) *http.ServeMux {
```
Comment (nfisher) [inline zpages/zpages.go]: Since it's returning a `*ServeMux` and there's no '/' specified on the mux it's probably ok to drop the mux parameter and create it in the method. `/public` is a fs folder so additional assets could be added at run-time and (famous last words) hopefully no one would want sub-paths under `/tracez` and `/rpcz`.

What are peoples thoughts on two functions?

1. `func zpages.NewPrefixed(pathPrefix string) *ServeMux`
2. `func zpages.New() *ServeMux`

The second basically calls the first with "" and would behaviour similar to the original `zpages.Handler`.

File: examples/grpc/helloworld_server/main.go [hunk 7]
```diff
@@ -47,10 +47,12 @@ func (s *server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloRe
 }
 
 func main() {
+	// Start z-Pages server.
 	go func() {
-		http.Handle("/debug/", http.StripPrefix("/debug", zpages.Handler))
-		log.Fatal(http.ListenAndServe(":8081", nil))
+		mux := zpages.AppendHandlers("/debug", http.NewServeMux())
+		log.Fatal(http.ListenAndServe("127.0.0.1:8081", mux))
```
Comment (nfisher) [inline examples/grpc/helloworld_server/main.go]: Like the change to 127.0.0.1 as it doesn't trigger FW notifications on OS X. 👍

File: zpages/zpages.go (hunk 6 above)
Comment (ramonza) [reply zpages/zpages.go]: I think constructing our own ServeMux reduces the flexibility, in particular to add to the DefaultServeMux. Renamed the method to Handle and no longer return the ServeMux to avoid ambiguity. 

I don't think we need a separate non-prefixed call. The user can just pass an empty string explicitly, I don't think it's worth adding a convenience method because I would think that the prefixed form would actually be more common in practice.

**Comment by ramonza** [comment]
@rakyll are you ok with the new interface?

**Comment by rakyll** [comment]
LGTM
